README.md
# flutter_speedometer

[Flutter](https://flutter.io) Speedometer widget package

![](flutter_speedometer_2.png)

### Getting Started

In order to use this package, do import
```dart
import 'package:flutter_speedometer/flutter_speedometer.dart';
```

Basic implementation can be done like below code:
```dart
import 'package:flutter/material.dart';
import 'package:flutter_speedometer/flutter_speedometer.dart';

void main() {
  runApp(
    Center(
      child: Speedometer(
        size: 250,
        minValue: 0,
        maxValue: 180,
        currentValue: 76,
        warningValue: 150,
        displayText: 'mph',
      ),
    ),
  );
}
```

### Example App
You can find more examples from [Example App](example)


### API
In this table, you can find all attributes provided by this package:

| Attribute           | Default value                     | Description |
| ------------------- | --------------------------------- | ----------- |
| size                | 200                               | Min value to be displayed   |
| minValue            | 0                                 | Min value to be displayed   |
| maxValue            | 100                               | Max value to be displayed   |
| currentValue        | 0                                 | Set the current value       |
| warningValue        | 80                                | Set the current value       |

### Objects
```dart
class Speedometer {

    final double size;
    final int minValue;
    final int maxValue;
    final int currentValue;
    final int warningValue;
    final Color backgroundColor;
    final Color meterColor;
    final Color warningColor;
    final Color kimColor;
    final TextStyle displayNumericStyle;
    final String displayText;
    final TextStyle displayTextStyle;
}
 ```


### Feedback

Feel free to [leave any feedback](https://github.com/ltdangkhoa/Flutter-Speedometer/issues) for helping support this package 🍻 


CHANGELOG.md
## [1.0.2] - 2019-08-07

* Change description

## [1.0.1] - 2019-08-07

* Fixed bugs

## [1.0.0] - 2019-07-11

* First release

## [0.0.1] - 2019-07-09

* Initial raw package


test/flutter_speedometer_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_speedometer/flutter_speedometer.dart';

void main() {
  test('adds one to input values', () {
    final speedometer = Speedometer();
  });
}


lib/src/painter.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';

class ArcPainter extends CustomPainter {
  ArcPainter(
      {this.startAngle = 0, this.sweepAngle = 0, this.color = Colors.white});
  final double startAngle;
  final double sweepAngle;
  final Color color;
  @override
  void paint(Canvas canvas, Size size) {
    final rect = Rect.fromLTRB(size.width * 0.05, size.width * 0.05,
        size.width * 0.95, size.height * 0.95);
    final startAngle = math.pi / 12 * this.startAngle;
    final sweepAngle = math.pi / 12 * this.sweepAngle;
    final useCenter = false;
    final paint = Paint()
      ..color = this.color
      ..style = PaintingStyle.stroke
      ..strokeWidth = size.width * 0.1;
    canvas.drawArc(rect, startAngle, sweepAngle, useCenter, paint);
  }

  @override
  bool shouldRepaint(CustomPainter old) {
    return false;
  }
}

class TriangleClipper extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    final path = Path();
    path.moveTo(size.width * 0.5, 0.0);
    path.lineTo(size.width, size.height);
    path.lineTo(0.0, size.height);
    path.close();
    return path;
  }

  @override
  bool shouldReclip(TriangleClipper oldClipper) => false;
}

class KimClipper extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    final path = Path();
    path.moveTo(size.width * 0.5, size.width * 0.5);
    path.lineTo(size.width * 0.5 + size.width / 30, size.height * 0.5);
    path.lineTo(size.width * 0.5 + 1, size.height - size.width / 30);
    path.lineTo(size.width * 0.5 - 1, size.height - size.width / 30);
    path.lineTo(size.width * 0.5 - size.width / 30, size.height * 0.5);
    path.close();
    return path;
  }

  @override
  bool shouldReclip(KimClipper oldClipper) => false;
}


example/lib/main.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import 'package:flutter_speedometer/flutter_speedometer.dart';

void main() {
  runApp(
    Center(
      child: Directionality(
        textDirection: TextDirection.ltr,
        child: Container(
          child: Flex(
            direction: Axis.vertical,
            children: <Widget>[
              Spacer(),
              Speedometer(
                size: 100,
                minValue: -10,
                maxValue: 50,
                currentValue: 30,
                warningValue: 40,
                backgroundColor: Colors.black,
                meterColor: Colors.lightBlueAccent,
                warningColor: Colors.orange,
                kimColor: Colors.purpleAccent,
                displayNumericStyle: TextStyle(
                    fontFamily: 'Digital-Display',
                    color: Colors.white,
                    fontSize: 18),
                displayText: '°C',
                displayTextStyle: TextStyle(color: Colors.white, fontSize: 8),
              ),
              Spacer(),
              Speedometer(
                size: 300,
                minValue: 0,
                maxValue: 180,
                currentValue: 58,
                warningValue: 90,
                backgroundColor: Colors.black,
                meterColor: Colors.green,
                warningColor: Colors.orange,
                kimColor: Colors.white,
                displayNumericStyle: TextStyle(
                    fontFamily: 'Digital-Display',
                    color: Colors.white,
                    fontSize: 40),
                displayText: 'km/h',
                displayTextStyle: TextStyle(color: Colors.white, fontSize: 15),
              ),
              Spacer(),
              Speedometer(),
              Spacer(),
            ],
          ),
        ),
      ),
    ),
  );
}


lib/flutter_speedometer.dart
library flutter_speedometer;

import 'dart:math' as math;
import 'package:flutter/material.dart';
import 'src/painter.dart';

class Speedometer extends StatefulWidget {
  Speedometer(
      {Key key,
      this.size = 200,
      this.minValue = 0,
      this.maxValue = 100,
      this.currentValue = 0,
      this.warningValue = 80,
      this.backgroundColor = Colors.black,
      this.meterColor = Colors.lightGreenAccent,
      this.warningColor = Colors.redAccent,
      this.kimColor = Colors.white,
      this.displayNumericStyle,
      this.displayText = '',
      this.displayTextStyle})
      : super(key: key);
  final double size;
  final int minValue;
  final int maxValue;
  final int currentValue;
  final int warningValue;
  final Color backgroundColor;
  final Color meterColor;
  final Color warningColor;
  final Color kimColor;
  final TextStyle displayNumericStyle;
  final String displayText;
  final TextStyle displayTextStyle;
  @override
  _SpeedometerState createState() => _SpeedometerState();
}

class _SpeedometerState extends State<Speedometer> {
  @override
  Widget build(BuildContext context) {
    double _size = widget.size;
    int _minValue = widget.minValue;
    int _maxValue = widget.maxValue;
    int _currentValue = widget.currentValue;
    int _warningValue = widget.warningValue;
    double startAngle = 3.0;
    double endAngle = 21.0;

    double _kimAngle = 0;
    if (_minValue <= _currentValue && _currentValue <= _maxValue) {
      _kimAngle = (((_currentValue - _minValue) * (endAngle - startAngle)) /
              (_maxValue - _minValue)) +
          startAngle;
    } else if (_currentValue < _minValue) {
      _kimAngle = startAngle;
    } else if (_currentValue > _maxValue) {
      _kimAngle = endAngle;
    }

    double startAngle2 = 0.0;
    double endAngle2 = 18.0;
    double _warningAngle = endAngle2;
    if (_minValue <= _warningValue && _warningValue <= _maxValue) {
      _warningAngle =
          (((_warningValue - _minValue) * (endAngle2 - startAngle2)) /
                  (_maxValue - _minValue)) +
              startAngle2;
    }
    return Container(
      color: widget.backgroundColor,
      child: Center(
        child: Directionality(
          textDirection: TextDirection.ltr,
          child: Container(
            width: _size,
            height: _size,
            child: Stack(
              children: <Widget>[
                Container(
                  alignment: Alignment.center,
                  padding: EdgeInsets.all(_size * 0.075),
                  child: Stack(children: <Widget>[
                    Container(
                      alignment: Alignment.center,
                      child: Container(
                        alignment: Alignment.center,
                        width: _size,
                        height: _size,
                        decoration: new BoxDecoration(
                          color: widget.backgroundColor,
                          boxShadow: [
                            new BoxShadow(
                                color: widget.kimColor,
                                blurRadius: 8.0,
                                spreadRadius: 4.0)
                          ],
                          shape: BoxShape.circle,
                        ),
                      ),
                    ),
                    CustomPaint(
                      size: Size(_size, _size),
                      painter: ArcPainter(
                          startAngle: 9,
                          sweepAngle: 18,
                          color: widget.warningColor),
                    ),
                    CustomPaint(
                      size: Size(_size, _size),
                      painter: ArcPainter(
                          startAngle: 9,
                          sweepAngle: _warningAngle,
                          color: widget.meterColor),
                    ),
                  ]),
                ),
                Container(
                  alignment: Alignment.bottomCenter,
                  child: ClipPath(
                    clipper: TriangleClipper(),
                    child: Container(
                      width: _size,
                      height: _size * 0.5,
                      color: widget.backgroundColor,
                    ),
                  ),
                ),
                Container(
                  alignment: Alignment.center,
                  child: Container(
                    width: _size * 0.1,
                    height: _size * 0.1,
                    alignment: Alignment.center,
                    decoration: new BoxDecoration(
                      color: widget.kimColor,
                      boxShadow: [
                        new BoxShadow(
                            color: widget.meterColor,
                            blurRadius: 10.0,
                            spreadRadius: 5.0)
                      ],
                      shape: BoxShape.circle,
                    ),
                  ),
                ),
                Container(
                  alignment: Alignment.center,
                  child: Transform.rotate(
                    angle: math.pi / 12 * _kimAngle,
                    child: ClipPath(
                      clipper: KimClipper(),
                      child: Container(
                        width: _size * 0.9,
                        height: _size * 0.9,
                        color: widget.kimColor,
                      ),
                    ),
                  ),
                ),
                Container(
                  alignment: Alignment.bottomCenter,
                  child: Padding(
                    padding: EdgeInsets.fromLTRB(0, 0, 0, 0),
                    child: Text(
                      widget.displayText,
                      style: widget.displayTextStyle,
                    ),
                  ),
                ),
                Container(
                  alignment: Alignment.bottomCenter,
                  child: Padding(
                    padding: EdgeInsets.fromLTRB(0, 0, 0, _size * 0.1),
                    child: Text(
                      widget.currentValue.toString(),
                      style: widget.displayNumericStyle,
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
